From eb70d2625f483a416488434b36a3defe30ad017c Mon Sep 17 00:00:00 2001
From: Yi Zhao <yi.zhao@windriver.com>
Date: Tue, 28 May 2024 15:06:06 +0800
Subject: [PATCH 2/2] policy/modules/system/selinuxutil: allow newrole to
 search

 faillock dir

Allow newrole to search the /run/faillock directory, otherwise the
faillock mechanism will not work for neworle.

Fixes:
avc: denied { search } for pid=508 comm="newrole" name="faillock"
dev="tmpfs" ino=582 scontext=root:sysadm_r:newrole_t:s0-s15:c0.c1023
tcontext=system_u:object_r:faillog_t:s0 tclass=dir permissive=0

Upstream-Status: Inappropriate [embedded specific]

Signed-off-by: Yi Zhao <yi.zhao@windriver.com>
---
 policy/modules/system/authlogin.if   | 18 ++++++++++++++++++
 policy/modules/system/selinuxutil.te |  1 +
 2 files changed, 19 insertions(+)

diff --git a/policy/modules/system/authlogin.if b/policy/modules/system/authlogin.if
index 62e0e0585..a68a8f2b2 100644
--- a/policy/modules/system/authlogin.if
+++ b/policy/modules/system/authlogin.if
@@ -956,6 +956,24 @@ interface(`auth_setattr_faillog_files',`
 	setattr_files_pattern($1, faillog_t, faillog_t)
 ')
 
+########################################
+## <summary>
+##	Search faillock dirs (/run/faillock).
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`auth_search_faillog_dirs',`
+	gen_require(`
+		type faillog_t;
+	')
+
+	allow $1 faillog_t:dir search_dir_perms;
+')
+
 #######################################
 ## <summary>
 ##	Read the last logins log.
diff --git a/policy/modules/system/selinuxutil.te b/policy/modules/system/selinuxutil.te
index 7770132d8..74b26bbd7 100644
--- a/policy/modules/system/selinuxutil.te
+++ b/policy/modules/system/selinuxutil.te
@@ -297,6 +297,7 @@ auth_run_upd_passwd(newrole_t, newrole_roles)
 auth_rw_faillog(newrole_t)
 auth_search_faillog(newrole_t)
 auth_read_shadow(newrole_t)
+auth_search_faillog_dirs(newrole_t)
 
 # Write to utmp.
 init_rw_utmp(newrole_t)
-- 
2.42.0

